### PR TITLE
Explorer: move deallocation to `StepCleanUp`

### PR DESCRIPTION
**Pull request by Pixep**
Move deallocation logic to `StepCleanUp` to have all necessary cleanup actions in the same place. Currently this means `DestroyAction` and `heap_.Deallocate`.

### CODE REVIEW

File: explorer/interpreter/action.h [hunk 1]
```diff
@@ -42,8 +42,7 @@ class RuntimeScope {
   RuntimeScope(RuntimeScope&&) noexcept;
   auto operator=(RuntimeScope&&) noexcept -> RuntimeScope&;
 
-  // Deallocates any allocations in this scope from `heap`.
-  ~RuntimeScope();
+  ~RuntimeScope() = default;
```
Comment (jonmeow) [inline explorer/interpreter/action.h]:
```suggestion
```

Can this be deleted? It looks like maybe it doesn't need to be explicit?

File: explorer/interpreter/interpreter.cpp [hunk 2]
```diff
@@ -2204,17 +2204,22 @@ auto Interpreter::StepDestroy() -> ErrorOr<Success> {
...
+  const Action& act = todo_.CurrentAction();
+  const auto& cleanup = cast<CleanUpAction>(act);
+  if (act.pos() < cleanup.allocations_count() * 2) {
+    const size_t alloc_index = cleanup.allocations_count() - act.pos() / 2 - 1;
+    auto allocation = act.scope()->allocations()[alloc_index];
+    if (act.pos() % 2 == 0) {
+      auto* lvalue = arena_->New<LValue>(Address(allocation));
+      auto value =
+          heap_.Read(lvalue->address(), SourceLocation("destructor", 1));
+      // Step over uninitialized values
```
Comment (jonmeow) [inline explorer/interpreter/interpreter.cpp]:
```suggestion
      // Step over uninitialized values.
```

Minor punctuation, only suggesting this given my other suggestion.

**Review by jonmeow — APPROVED** [review]

File: explorer/interpreter/action.h (hunk 1 above)
Comment (Pixep) [reply explorer/interpreter/action.h]: Can do yes.

I preferred to keep it given we should be defining the big 3, clarifying that we want the default implementation on purpose. i.e. we didn't forget to define it.

Comment (jonmeow) [reply explorer/interpreter/action.h]: FWIW my leaning would still be to delete, I don't have the "forgot to define a destructor" instinct [to the contrary, I see a destructor and think "why is this required"]. But I did approve, so do as you like.

Comment (Pixep) [reply explorer/interpreter/action.h]: Done

**Pull request merged**